### PR TITLE
Allow modifying underlying data from non-const ArrayView

### LINKED ISSUE #732: Semantics of ArrayView and const
Should we be able to modify the data within a const ArrayView?

We should probably follow the semantics of std::span

### PR DESCRIPTION
**Pull request by kanye-quest**
# Summary

- Resolves #732 
- Enables modifying data pointed to by `ArrayView<T>` when `T` is not const
  - Removes requirement for marking lambdas as `mutable` in order to use `ArrayView`
- Adds `ArrayTraits<T>` class for passing static information to `ArrayBase` CRTP class

### CODE REVIEW

File: src/axom/core/Array.hpp [hunk 1]
```diff
@@ -32,6 +32,12 @@ struct Uninitialized
 template <typename T, int DIM, MemorySpace SPACE>
 class Array;
 
+template <typename T, int DIM, MemorySpace SPACE>
+struct ArrayTraits<Array<T, DIM, SPACE>>
```
[the remark this answers is not in the document]
Comment (joshessman-llnl) [reply src/axom/core/Array.hpp:36]: Minor/thought: there's also the approach of defining a `constexpr static bool is_view` in `ArrayView` and `Array`, but I think your approach is probably better especially if we intend on defining other arraytraits members, plus yours doesn't "pollute" Array

File: src/axom/core/ArrayBase.hpp [hunk 2]
```diff
@@ -83,11 +86,27 @@ bool operator!=(const ArrayBase<T1, DIM, LArrayType>& lhs,
...
+
 public:
+  /* If ArrayType is an ArrayView, we use shallow-const semantics, akin to
+   * std::span; a const ArrayView will still allow for mutating the underlying
+   * pointed-to data.
+   *
+   * If ArrayType is an Array, we use deep-const semantics, akin to std::vector;
+   * a const Array will prevent modifications of the underlying Array data.
+   */
+  using ConstT = typename std::conditional<IsArrayView, T, const T>::type;
```
Comment (joshessman-llnl) [inline src/axom/core/ArrayBase.hpp:108]: :+1: good docs!

Minor: maybe a more descriptive name, perhaps `true_const_t` or something? Not totally sure

File: src/axom/core/ArrayView.hpp [hunk 3]
```diff
@@ -162,6 +144,10 @@ ArrayView<T, DIM, SPACE>::ArrayView(T* data, Args... args)
 {
   static_assert(sizeof...(Args) == DIM,
                 "Array size must match number of dimensions");
+#ifdef AXOM_DEVICE_CODE
```
Comment (joshessman-llnl) [inline src/axom/core/ArrayView.hpp:147]: Note that this is subject to #734

**Review by joshessman-llnl — APPROVED** [review]
Thanks @kanye-quest !

File: src/axom/core/ArrayView.hpp (hunk 3 above)
Comment (joshessman-llnl) [reply src/axom/core/ArrayView.hpp:147]: Ah this looks like it's causing a CI failure as well

File: src/axom/core/ArrayView.hpp [hunk 4]
```diff
@@ -16,6 +16,12 @@ namespace axom
 template <typename T, int DIM, MemorySpace SPACE>
 class ArrayView;
 
+template <typename T, int DIM, MemorySpace SPACE>
+struct ArrayTraits<ArrayView<T, DIM, SPACE>>
+{
+  constexpr static bool IsView = true;
```
Comment (white238) [inline src/axom/core/ArrayView.hpp:22]: I agree with @joshessman-llnl .  I'm pretty sure we only use camel-casing for classes. I would prefer pot-hole casing.

@kennyweiss ?

File: src/axom/core/examples/core_containers.cpp [hunk 5]
```diff
@@ -287,13 +287,10 @@ void demoArrayDevice()
...
 
-  // Declare the lambda mutable so our copy of C_view (captured by value) is mutable
-  axom::for_all<axom::CUDA_EXEC<1>>(
-    0,
-    N,
-    [=] AXOM_HOST_DEVICE(axom::IndexType i) mutable {
-      C_view[i] = A_view[i] + B_view[i] + 1;
-    });
+  // Write to the underlying array through C_view, which is captured by value
+  axom::for_all<axom::CUDA_EXEC<1>>(0, N, [=] AXOM_HOST_DEVICE(axom::IndexType i) {
```
Comment (white238) [inline src/axom/core/examples/core_containers.cpp:291]: It's much nicer not to have to remember to put `mutable` here.  Thanks @kanye-quest !

**Review by white238 — APPROVED** [review]
Approved pending response/fixing of comments.

File: src/axom/core/docs/sphinx/core_containers.rst [hunk 6]
```diff
@@ -241,8 +241,6 @@ via a lambda:
    :end-before: _array_w_raja_end
    :language: C++
 
-.. note:: We need to mark the lambda as `mutable` if we want to modify the array (array view) data.
```
Comment (kennyweiss) [inline src/axom/core/docs/sphinx/core_containers.rst:244]: 👍

**Review by kennyweiss — APPROVED** [review]
Thanks @kanye-quest !

File: src/axom/core/ArrayView.hpp (hunk 4 above)
Comment (rhornung67) [reply src/axom/core/ArrayView.hpp:22]: Yes, we use camelCaseForClassesAndMethods. Typically, we use pot-hole case for variable names.

File: src/axom/core/ArrayBase.hpp (hunk 2 above)
Comment (kanye-quest) [reply src/axom/core/ArrayBase.hpp]: Agreed - I renamed our conditional type `RealConstT`

**Pull request merged**